Sources/BrowserServicesKit/FeatureFlagger/ExperimentCohortsManager.swift

```diff
@@ -0,0 +1,146 @@
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public struct ExperimentSubfeature {
+    let parentID: ParentFeatureID
+    let subfeatureID: SubfeatureID
+    let cohorts: [PrivacyConfigurationData.Cohort]
+}
+
+public typealias CohortID = String
+public typealias SubfeatureID = String
+public typealias ParentFeatureID = String
+
+public struct ExperimentData: Codable, Equatable {
+    public let parentID: String
+    public let cohortID: String
+    public let enrollmentDate: Date
+}
+
+public typealias Experiments = [String: ExperimentData]
+
+public protocol ExperimentCohortsManaging {
+    /// Retrieves all the experiments a user is enrolled in
+    var experiments: Experiments? { get }
+
+    /// Resolves the cohort for a given experiment subfeature.
+    ///
+    /// This method determines whether the user is currently assigned to a valid cohort
+    /// for the specified experiment. If the assigned cohort is valid (i.e., it matches
+    /// one of the experiment's defined cohorts), the method returns the assigned cohort.
+    /// Otherwise, the invalid cohort is removed, and a new cohort is assigned if
+    /// `isAssignCohortEnabled` is `true`.
+    ///
+    /// - Parameters:
+    ///   - experiment: The `ExperimentSubfeature` representing the experiment and its associated cohorts.
+    ///   - isAssignCohortEnabled: A Boolean value indicating whether cohort assignment is allowed
+    ///     if the user is not already assigned to a valid cohort.
+    ///
+    /// - Returns: The valid `CohortID` assigned to the user for the experiment, or `nil`
+    ///   if no valid cohort exists and `isAssignCohortEnabled` is `false`.
+    ///
+    /// - Behavior:
+    ///   1. Retrieves the currently assigned cohort for the experiment using the `subfeatureID`.
+    ///   2. Validates if the assigned cohort exists within the experiment's cohort list:
+    ///      - If valid, the assigned cohort is returned.
+    ///      - If invalid, the cohort is removed from storage.
+    ///   3. If cohort assignment is enabled (`isAssignCohortEnabled` is `true`), a new cohort
+    ///      is assigned based on the experiment's cohort weights and saved in storage.
+    ///      - Cohort assignment is probabilistic, determined by the cohort weights.
+    ///
+    func resolveCohort(for experiment: ExperimentSubfeature, isAssignCohortEnabled: Bool) -> CohortID?
```

Comment:
sorry for getting back to this, but after reading this piece of documentation I think allowCohortReassignment might be a clearer name, but leaving it up to you, no strong feelings about it